core/src/main/scala/scalaz/Codensity.scala

```diff
@@ -20,6 +31,23 @@ abstract class Codensity[F[_], A] { self =>
 }
 
 object Codensity extends CodensityInstances {
+
+  private[Codensity] abstract class IntrusiveEphemeralStream[F[_], A]
+      extends Codensity[F, A] {
+    def tail: Codensity[F, A]
+    @tailrec final def last: Codensity[F, A] = {
+      tail match {
+        case stream: IntrusiveEphemeralStream[F, A] =>
```

Comment:
We cannot let `IntrusiveEphemeralStream` be generic because this type check must be specialized for each type